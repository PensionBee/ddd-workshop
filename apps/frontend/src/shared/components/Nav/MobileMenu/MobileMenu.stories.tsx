import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { LIST_WITH_MIXED_ITEMS } from "../Nav.content";
import { MemoryRouter } from "react-router-dom";
import { Box } from "@shared/components/Box";
import type { MobileMenuProps } from "./MobileMenu";
import { MobileMenu } from "./MobileMenu";

export default {
  title: "Shared/Menus/Mobile Menu",
  component: MobileMenu,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof MobileMenu>;

const MobileMenuTemplate: ComponentStory<typeof MobileMenu> = (
  args: MobileMenuProps
) => {
  return (
    <MemoryRouter initialEntries={[{ pathname: "/example" }]}>
      <Box d="flex" direction="column" w={320}>
        <MobileMenu {...args} />
      </Box>
    </MemoryRouter>
  );
};

export const Default = MobileMenuTemplate.bind({});
Default.args = {
  subList: LIST_WITH_MIXED_ITEMS,
};
