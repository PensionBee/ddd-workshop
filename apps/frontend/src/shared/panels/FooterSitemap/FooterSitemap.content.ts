export const SERVICE_LINKS = [
  { href: "/how-it-works", linkText: "How it works" },
  { href: "/combine", linkText: "Combine" },
  { href: "/contribute", linkText: "Contribute" },
  { href: "/pension-withdrawal", linkText: "Withdraw" },
  { href: "/plans", linkText: "Plans" },
  { href: "/self-employed-pension", linkText: "Self-employed" },
  { href: "/fees", linkText: "Fees" },
];

export const ABOUT_LINKS = [
  { href: "/our-vision", linkText: "Vision" },
  { href: "/our-values", linkText: "Values" },
  { href: "/esg", linkText: "ESG" },
  { href: "/our-history", linkText: "History" },
  { href: "/reviews", linkText: "Reviews" },
  { href: "/press", linkText: "Press" },
  { href: "/careers", linkText: "Careers" },
];

export const KNOWLEDGE_LINKS = [
  { href: "/pensions-explained", linkText: "Pensions explained" },
  { href: "/pension-academy-videos", linkText: "Pension Academy videos" },
  { href: "/blog", linkText: "Blog" },
  { href: "/podcast", linkText: "Pension Confident Podcast" },
  { href: "/pension-calculator", linkText: "Pension calculator" },
  { href: "/drawdown-calculator", linkText: "Drawdown calculator" },
  { href: "/pension-landscape", linkText: "Pension landscape" },
];

export const INVESTOR_LINKS = [
  { href: "/investor-relations", linkText: "Overview" },
  {
    href: "/investor-relations/business-model-and-strategy",
    linkText: "Business model & strategy",
  },
  { href: "/investor-relations/esg", linkText: "ESG document hub" },
  { href: "/investor-relations/leadership", linkText: "Leadership" },
  { href: "/investor-relations/regulatory-news", linkText: "Regulatory news" },
  {
    href: "/investor-relations/results-and-reports",
    linkText: "Results and reports",
  },
  {
    href: "/investor-relations/financial-calendar",
    linkText: "Financial calendar",
  },
  {
    href: "/investor-relations/share-price-and-analyst-coverage",
    linkText: "Share price and analyst coverage",
  },
  {
    href: "/investor-relations/annual-general-meeting",
    linkText: "Annual General Meeting",
  },
  { href: "/investor-relations/ipo-centre", linkText: "IPO centre" },
  { href: "/investor-relations/contact-us", linkText: "Contact us" },
];

export const HELP_LINKS = [
  { href: "/contact-us", linkText: "Contact us" },
  { href: "/faq", linkText: "FAQs" },
  { href: "/terms", linkText: "Terms" },
  { href: "/privacy-policy", linkText: "Privacy policy" },
];

export const FOOTER_MENU_SECTIONS = {
  "OUR SERVICE": SERVICE_LINKS,
  "ABOUT US": ABOUT_LINKS,
  KNOWLEDGE: KNOWLEDGE_LINKS,
  "INVESTOR RELATIONS": INVESTOR_LINKS,
  "HELP & FAQS": HELP_LINKS,
} as const;
