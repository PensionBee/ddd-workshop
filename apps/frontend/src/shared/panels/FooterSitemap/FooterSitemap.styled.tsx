import { LargeSystemIcon } from "@shared/components/Svg";
import styled from "styled-components";

export const StyledCaret = styled(LargeSystemIcon)`
  position: absolute;
  right: 2rem;
  top: 50%;
  transform: translateY(-50%) rotate(-180deg);
  transition: transform 250ms ease;
  [data-state="open"] & {
    transform: translateY(-50%) rotate(0deg);
  }
`;

export const StyledAccordionButton = styled.button`
  cursor: pointer;
  position: relative;
  width: 100%;
  padding-top: 20px;
  padding-bottom: 20px;
  text-align: center;
`;
