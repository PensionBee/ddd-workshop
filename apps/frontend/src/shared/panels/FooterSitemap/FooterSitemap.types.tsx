import type { FOOTER_MENU_SECTIONS } from "./FooterSitemap.content";
import type { Link } from "@shared/components/Link";

export type Menu = {
  sections: typeof FOOTER_MENU_SECTIONS;
};

export type MenuLink = {
  linkText: string;
  href: React.ComponentProps<typeof Link>["href"];
  target?: React.ComponentProps<typeof Link>["target"];
};

export type MenuList = {
  title: string;
  list: MenuLink[];
};

export type MobileMenuList = MenuList & {
  borderTop?: boolean;
  onClick?: () => void;
  open?: boolean;
};
