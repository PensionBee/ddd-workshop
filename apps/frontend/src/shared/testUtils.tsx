import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
