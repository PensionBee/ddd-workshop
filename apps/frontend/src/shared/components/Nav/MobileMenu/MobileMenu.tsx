import type { NavMenuItem } from "../Nav.types";
import { StyledContentList } from "./MobileMenu.styled";
import { MobileMenuSubList } from "./MobileMenuSubList/MobileMenuSubList";

export type MobileMenuProps = {
  subList: NavMenuItem["subList"];
};

export const MobileMenu: React.FC<MobileMenuProps> = ({ subList }) => {
  return (
    <StyledContentList>
      <MobileMenuSubList subList={subList} />
    </StyledContentList>
  );
};

export type MobileMainMenuProps = {
  list: NavMenuItem[];
};

export const MobileMainMenu: React.FC<MobileMainMenuProps> = ({ list }) => (
  <MobileMenu subList={list} />
);
