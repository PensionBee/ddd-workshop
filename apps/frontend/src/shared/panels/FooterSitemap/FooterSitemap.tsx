import { useMediaQueryLg } from "@shared/hooks/useMediaQuery/useMediaQuery";
import { DesktopMenu } from "./components/DesktopMenu";
import { MobileMenu } from "./components/MobileMenu";
import { FOOTER_MENU_SECTIONS } from "./FooterSitemap.content";

export const FooterSitemap: React.FC = (props) => {
  const isDesktop = useMediaQueryLg();
  if (isDesktop) {
    return (
      <DesktopMenu
        data-rc="FooterSitemap"
        data-testid="DesktopMenu"
        sections={FOOTER_MENU_SECTIONS}
        {...props}
      />
    );
  }
  return (
    <MobileMenu
      data-rc="FooterSitemap"
      data-testid="MobileMenu"
      sections={FOOTER_MENU_SECTIONS}
      {...props}
    />
  );
};
