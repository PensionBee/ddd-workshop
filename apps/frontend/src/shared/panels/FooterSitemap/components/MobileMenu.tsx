import { GridContainer, GridItem, GridRow } from "@shared/components/Grid";
import { BodySmallL, Category } from "@shared/components/Typography";
import { useState } from "react";
import { Link } from "@shared/components/Link";
import { Box } from "@shared/components/Box";
import { StyledAccordionButton, StyledCaret } from "../FooterSitemap.styled";
import type { MenuLink, MobileMenuList, Menu } from "../FooterSitemap.types";

export const MobileLinkItem: React.FC<MenuLink> = ({
  linkText,
  href,
  target,
  ...props
}) => {
  return (
    <Box as="li" d="flex" pt={10} pb={10} justify="center" {...props}>
      <BodySmallL align="center">
        <Link href={href} target={target} color="black">
          {linkText}
        </Link>
      </BodySmallL>
    </Box>
  );
};

export const MobileAccordion: React.FC<MobileMenuList> = ({
  title,
  list,
  borderTop,
  onClick,
  open,
  ...props
}) => {
  return (
    <GridRow data-state={open ? "open" : "closed"} {...props}>
      <GridItem align="center">
        {borderTop && <Box borderColor="grey4" borderTop={1} w="100%" />}
        <StyledAccordionButton onClick={onClick}>
          <Category>{title}</Category>
          <StyledCaret icon="ChevronYellow" aria-hidden />
        </StyledAccordionButton>
        {open && (
          <Box as="ul" mb={24}>
            {list.map((props) => (
              <MobileLinkItem key={props.linkText} {...props} />
            ))}
          </Box>
        )}
        <Box borderColor="grey4" borderBottom={1} w="100%" />
      </GridItem>
    </GridRow>
  );
};

export const MobileMenu: React.FC<Menu> = ({ sections, ...props }) => {
  const [open, setOpen] = useState<string | null>(null);

  const handleSetOpen = (title: string) => {
    return setOpen((open) => {
      if (title === open) return null;
      return title;
    });
  };

  return (
    <GridContainer pt={{ xs: 40 }} {...props}>
      {Object.entries(sections).map(([key, value], index) => (
        <MobileAccordion
          key={key}
          title={key}
          list={value}
          onClick={() => handleSetOpen(key)}
          open={open === key}
          borderTop={index === 0}
        />
      ))}
    </GridContainer>
  );
};
