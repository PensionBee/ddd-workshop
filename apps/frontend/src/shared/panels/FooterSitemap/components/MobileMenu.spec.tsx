import { act, fireEvent, queryByText, render } from "@shared/testUtils";
import { FOOTER_MENU_SECTIONS } from "../FooterSitemap.content";
import { MobileMenu, MobileLinkItem, MobileAccordion } from "./MobileMenu";

const CLICK_EVENT = new MouseEvent("click", {
  bubbles: true,
  cancelable: true,
});

describe("MobileMenu", () => {
  const TEST_ID = "MobileMenu";
  it("should render", () => {
    const { getByTestId, asFragment } = render(
      <MobileMenu data-testid={TEST_ID} sections={FOOTER_MENU_SECTIONS} />
    );
    expect(getByTestId(TEST_ID)).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });

  it("should open and close accordions", async () => {
    const { getByTestId, asFragment } = render(
      <MobileMenu data-testid={TEST_ID} sections={FOOTER_MENU_SECTIONS} />
    );
    const accordion = getByTestId(TEST_ID).querySelector("[data-state]");
    expect(accordion).toBeInTheDocument();
    const button = accordion && accordion.querySelector("button");
    expect(accordion).toHaveAttribute("data-state", "closed");
    act(() => button && fireEvent(button, CLICK_EVENT));
    expect(accordion).toHaveAttribute("data-state", "open");
    act(() => button && fireEvent(button, CLICK_EVENT));
    expect(accordion).toHaveAttribute("data-state", "closed");
    expect(asFragment).toMatchSnapshot();
  });
});

describe("MobileLinkItem", () => {
  const TEST_ID = "MobileLinkItem";
  it("should render", () => {
    const { getByTestId, asFragment } = render(
      <MobileLinkItem data-testid={TEST_ID} linkText={""} href={undefined} />
    );
    expect(getByTestId(TEST_ID)).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });
});

describe("MobileAccordion", () => {
  const TEST_ID = "MobileAccordion";
  const MENU_TITLE = "Menu Title";
  const EXAMPLE_LIST = [{ linkText: "Link 1", href: "/example" }];
  it("should render", () => {
    const { getByTestId, asFragment } = render(
      <MobileAccordion
        data-testid={TEST_ID}
        title={MENU_TITLE}
        list={EXAMPLE_LIST}
      />
    );
    const accordion = getByTestId(TEST_ID);
    expect(accordion).toBeInTheDocument();
    expect(accordion).toHaveAttribute("data-state", "closed");
    expect(
      queryByText(accordion, EXAMPLE_LIST[0].linkText)
    ).not.toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });

  it("should render with content when open", async () => {
    const { getByTestId, asFragment } = render(
      <MobileAccordion
        data-testid={TEST_ID}
        title={MENU_TITLE}
        list={EXAMPLE_LIST}
        open={true}
      />
    );
    const accordion = getByTestId(TEST_ID);
    expect(accordion).toBeInTheDocument();
    expect(accordion).toHaveAttribute("data-state", "open");
    expect(
      queryByText(accordion, EXAMPLE_LIST[0].linkText)
    ).toBeInTheDocument();
    expect(asFragment).toMatchSnapshot();
  });
});
