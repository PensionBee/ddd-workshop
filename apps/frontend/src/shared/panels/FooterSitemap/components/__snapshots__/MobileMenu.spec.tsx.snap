// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MobileAccordion should render 1`] = `[Function]`;

exports[`MobileAccordion should render with content when open 1`] = `[Function]`;

exports[`MobileLinkItem > should render 1`] = `[Function]`;

exports[`MobileLinkItem should render 1`] = `[Function]`;

exports[`MobileMenu should open and close accordions 1`] = `[Function]`;

exports[`MobileMenu should render 1`] = `[Function]`;
